Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/SOCKSEventsHandler.swift

```diff
@@ -0,0 +1,117 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOSOCKS
+
+final class SOCKSEventsHandler: ChannelInboundHandler, RemovableChannelHandler {
+    typealias InboundIn = NIOAny
+
+    enum State {
+        // transitions to channelActive or failed
+        case initialized
+        // transitions to socksEstablished or failed
+        case channelActive(Scheduled<Void>)
+        // final success state
+        case socksEstablished
+        // final success state
+        case failed(Error)
+    }
+
+    private var socksEstablishedPromise: EventLoopPromise<Void>?
+    var socksEstablishedFuture: EventLoopFuture<Void>! {
```

Comment:
SSWG rule

Reply:
Also I don't think this is actually safe as the underlying promise is a "real" optional.